File: src/calculate.rs
use super::parser::{Node, Operator};

fn traverse(node: &Node) -> f32 {
    use Node::{*};

    let (op, left, right) = match node {
        Op(op, left, right) => (op, left, right),
        Num(n) => return *n as f32
    };

    use Operator::{*};

    let left = traverse(left);
    let right = traverse(right);

    match op {
        Plus => left + right,
        Minus => left - right,
        Divide => left / right,
        Multiply => left * right
    }
}

pub fn calculate(node: &Node) -> f32 {
    traverse(node)
}

File: calculate.rb
require "./lexer.rb"
require "./parser.rb"

def traverse(node)
    if node.instance_of? Integer
        return node
    else
        left = traverse(node.left)
        right = traverse(node.right)

        if node.op == '+'
            return left + right
        elsif node.op == '-'
            return left - right
        elsif node.op == '/'
            return left / right
        else # node.op == '*''
            return left * right
        end
    end 
end

def calculate(input)
    tokens = tokenize(input)
    tree = parse(tokens)
    traverse(tree)
end